### PR TITLE
feat: deploy starknet account if CairoLib caller is not deployed

### PR DESCRIPTION
**Pull request by enitrat**
Time spent on this PR: 0.3d

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #<Issue number>

## What is the new behavior?

- Deploys a SN account for a caller of a Cairo contract if its account is not deployed on starknet already.
-
-

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1253)

### CODE REVIEW

File: src/backend/starknet.cairo [hunk 1]
```diff
@@ -183,42 +183,31 @@ namespace Internals {
...
-                starknet_address, self.valid_jumpdests_start, self.valid_jumpdests
-            );
-            return ();
+            tempvar syscall_ptr = syscall_ptr;
+            tempvar pedersen_ptr = pedersen_ptr;
+            tempvar range_check_ptr = range_check_ptr;
+        } else {
+            tempvar syscall_ptr = syscall_ptr;
+            tempvar pedersen_ptr = pedersen_ptr;
+            tempvar range_check_ptr = range_check_ptr;
```
Comment (ClementWalter) [inline src/backend/starknet.cairo:192]: if we update the signature of `Starknet.deploy` to not return the (useless) starknet address, we can save 3 steps here

```suggestion
        } else {
            tempvar syscall_ptr = syscall_ptr;
            tempvar pedersen_ptr = pedersen_ptr;
            tempvar range_check_ptr = range_check_ptr;
         }
         let syscall_ptr = cast([ap - 3], felt*);
         let pedersen_ptr = ...
```

File: src/kakarot/precompiles/kakarot_precompiles.cairo [hunk 2]
```diff
@@ -82,11 +83,22 @@ namespace KakarotPrecompiles {
...
+                let (deployed_contract) = Starknet.deploy(caller_address);
+                tempvar caller_starknet_address = deployed_contract;
+                tempvar syscall_ptr = syscall_ptr;
+                tempvar pedersen_ptr = pedersen_ptr;
+                tempvar range_check_ptr = range_check_ptr;
+            } else {
+                tempvar caller_starknet_address = caller_starknet_address;
+                tempvar syscall_ptr = syscall_ptr;
+                tempvar pedersen_ptr = pedersen_ptr;
+                tempvar range_check_ptr = range_check_ptr;
```
Comment (ClementWalter) [inline src/kakarot/precompiles/kakarot_precompiles.cairo:96]: ok, forget my comment above but here you can do
```suggestion
                // Deploy account
                Starknet.deploy(caller_address);
            } else {
                tempvar syscall_ptr = syscall_ptr;
                tempvar pedersen_ptr = pedersen_ptr;
                tempvar range_check_ptr = range_check_ptr;
                tempvar caller_starknet_address = caller_starknet_address;                
```

Comment (enitrat) [reply src/kakarot/precompiles/kakarot_precompiles.cairo:96]: I don't get it - how do you get the caller starknet address?

File: tests/utils/syscall_handler.py [hunk 3]
```diff
@@ -510,6 +511,61 @@ def send_message_to_l1(self, segments, syscall_ptr):
...
+    def patch_deploy(cls, value: callable):
+        """
+        Patch the deploy syscall with the value.
+
+        :param value: The value to patch with, a callable that will be called with the class hash,
+            the contract address salt, the constructor calldata and the deploy from zero flag.
+        """
+        cls.patches["deploy"] = value
+        yield
+        del cls.patches["deploy"]
```
Comment (ClementWalter) [inline tests/utils/syscall_handler.py:567]: if we use the same `patches` object, why not keeping one single `patch` decorator and doing `@SyscallHandler.patch("deploy", ...)`

maybe better to do `@SyscallHandler.patch("deploy_syscall", ...)`

File: src/kakarot/precompiles/kakarot_precompiles.cairo (hunk 2 above)
Comment (ClementWalter) [reply src/kakarot/precompiles/kakarot_precompiles.cairo:96]: the return data of the `Starknet.deploy` is in this order

```
syscall_ptr: felt*
pedersen_ptr: HashBuiltin*
range_check_ptr
account_address
```

so you don't need to tempvar, just use the same ordering

File: tests/utils/syscall_handler.py (hunk 3 above)
Comment (enitrat) [reply tests/utils/syscall_handler.py:567]: tbh I think having "patch_storage" "patch_call" "patch_deploy" would make more sense than a one-fits-all

Comment (enitrat) [reply tests/utils/syscall_handler.py:567]: I personnally prefer this way and would even create the `patch_storage` / `patch_call` distinction but it's not very useful refacto

Comment (ClementWalter) [reply tests/utils/syscall_handler.py:567]: it's debatable, the official lib ha one unique `patch` decorator

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**